<a href="https://colab.research.google.com/github/shadfdz/speechEmotionRecognition/blob/master/Speech_emotion_recognition_preprocess.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [23]:
import os

import librosa
import librosa.display
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import glob
import re
import torch
from google.colab import drive
from torch.utils.data import Dataset, DataLoader

drive.mount('/content/drive',force_remount=True)

Mounted at /content/drive


In [ ]:
!ls drive/MyDrive/speech_emotion_dataset/Crema

In [ ]:
# Create folders and group each emotion in the same folder

os.chdir('drive/MyDrive')
os.mkdir('processed_emotion_dataset')
os.chdir('processed_emotion_dataset')

emotion_cat = ['neutral', 'calm', 'happy', 'sad', 'angry', 'fearful', 'disgust', 'surprised']

for cat in emotion_cat:
  os.mkdir(cat)

In [ ]:
# show folder for each emotion
os.chdir(os.path.dirname(os.getcwd()))
!ls 

In [26]:
dir_path = 'drive/MyDrive/speech_emotion_dataset/'
list_ds = glob.glob(dir_path + "*") 

# Export Crema Dataset to Respective Emotion Folder

In [ ]:
# get crema file names
crema_file_list = os.listdir(list_ds[1] + "/")

In [31]:
# get file paths of target
emotion_folder_path_list = glob.glob('drive/MyDrive/processed_emotion_dataset/*')

In [32]:
# create dictionary for emotion and file path
emotion_folder_dict = {}
for cat_path in emotion_folder_path_list:
  emotion_folder_dict[cat_path.split('/')[3]] = cat_path

In [ ]:
# get crema file paths
crema_paths = glob.glob(emotion_folder_path_list)

['drive/MyDrive/speech_emotion_dataset/Savee',
 'drive/MyDrive/speech_emotion_dataset/Crema',
 'drive/MyDrive/speech_emotion_dataset/Tess',
 'drive/MyDrive/speech_emotion_dataset/Ravdess']

In [ ]:
# get file paths
crema_path_list = glob.glob(list_ds[1] + "/*")

In [ ]:
# dictionary to map to emotion folder
crema_emotions = {
        "SAD": "sad",
        "ANG": "angry",
        "DIS": "disgust",
        "FEA": "fearful",
        "HAP": "happy",
        "NEU": "neutral",
}

In [33]:
emotion_folder_dict

{'angry': 'drive/MyDrive/processed_emotion_dataset/angry',
 'calm': 'drive/MyDrive/processed_emotion_dataset/calm',
 'disgust': 'drive/MyDrive/processed_emotion_dataset/disgust',
 'fearful': 'drive/MyDrive/processed_emotion_dataset/fearful',
 'happy': 'drive/MyDrive/processed_emotion_dataset/happy',
 'neutral': 'drive/MyDrive/processed_emotion_dataset/neutral',
 'sad': 'drive/MyDrive/processed_emotion_dataset/sad',
 'surprised': 'drive/MyDrive/processed_emotion_dataset/surprised'}

In [ ]:
# sort crema to appropriate emotion category
for file in crema_path_list:
  os.system('cp ' + file + ' ' + emotion_folder_dict.get(crema_emotions.get(file.split('_')[4])))

# Export Savee Dataset to Respective Emotion Folder


In [35]:
# create dictionary for savee emotions
savee_emotions = {
    "a": "angry",
    "d": "disgust",
    "f": "fearful",
    "h": "happy",
    "n": "neutral",
    "sa": "sad",
    "su": "surprised",
}

In [36]:
# get path of files in savee folder as list
savee_path_list = glob.glob(list_ds[0] + "/*")

In [34]:
# copy files to respective emotion
for file in savee_path_list:
  os.system('cp ' + file + ' ' + emotion_folder_dict.get(savee_emotions.get(re.sub(r'\d+','',file.split('_')[3]).replace('.wav',''))))

# Export Ravdes Dataset to Respective Emotion Folder


In [37]:
# create emotion dictionary to map to folder
ravdess_emotions = {
    "01": "neutral",
    "02": "calm",
    "03": "happy",
    "04": "sad",
    "05": "angry",
    "06": "fearful",
    "07": "disgust",
    "08": "surprised",
}

In [38]:
list_ds[3]

'drive/MyDrive/speech_emotion_dataset/Ravdess'

In [ ]:
# list files
!ls drive/MyDrive/speech_emotion_dataset/Ravdess/audio_speech_actors_01-24/Actor_01

In [39]:
# get file paths as list
ravdes_path_list = glob.glob('drive/MyDrive/speech_emotion_dataset/Ravdess/audio_speech_actors_01-24/Actor_01' + '/*')

In [ ]:
len(ravdes_path_list)

60

In [40]:
# move file to respective emotional folder
for file in ravdes_path_list:
  os.system('cp ' + file + ' ' + emotion_folder_dict.get(ravdess_emotions.get(file.split('-')[3])))
  

# Export Tess Dataset to Respective Emotion Folder


In [41]:
list_ds[2]

'drive/MyDrive/speech_emotion_dataset/Tess'

In [ ]:
# list Tess folders
!ls drive/MyDrive/speech_emotion_dataset/Tess/

OAF_angry    OAF_neutral	    YAF_disgust  YAF_pleasant_surprised
OAF_disgust  OAF_Pleasant_surprise  YAF_fear	 YAF_sad
OAF_Fear     OAF_Sad		    YAF_happy
OAF_happy    YAF_angry		    YAF_neutral


In [42]:
# create dictionary map to emotion category
tess_emotions = {
    "neutral": "neutral",
    "happy": "happy",
    "sad": "sad",
    "angry": "angry",
    "Fear": "fearful",
    "disgust": "disgust",
    "surprised": "surprised",
    "surprise": "surprised",
    "Sad": "sad",
    "fear": "fearful"
}

In [43]:
tess_folder_list = glob.glob('drive/MyDrive/speech_emotion_dataset/Tess' + '/*')
tess_folder_list

['drive/MyDrive/speech_emotion_dataset/Tess/YAF_pleasant_surprised',
 'drive/MyDrive/speech_emotion_dataset/Tess/OAF_happy',
 'drive/MyDrive/speech_emotion_dataset/Tess/OAF_angry',
 'drive/MyDrive/speech_emotion_dataset/Tess/YAF_happy',
 'drive/MyDrive/speech_emotion_dataset/Tess/YAF_neutral',
 'drive/MyDrive/speech_emotion_dataset/Tess/YAF_disgust',
 'drive/MyDrive/speech_emotion_dataset/Tess/OAF_Fear',
 'drive/MyDrive/speech_emotion_dataset/Tess/YAF_sad',
 'drive/MyDrive/speech_emotion_dataset/Tess/OAF_Pleasant_surprise',
 'drive/MyDrive/speech_emotion_dataset/Tess/OAF_disgust',
 'drive/MyDrive/speech_emotion_dataset/Tess/OAF_neutral',
 'drive/MyDrive/speech_emotion_dataset/Tess/OAF_Sad',
 'drive/MyDrive/speech_emotion_dataset/Tess/YAF_fear',
 'drive/MyDrive/speech_emotion_dataset/Tess/YAF_angry']

In [44]:
# transfer files
for subfolder in tess_folder_list:
  os.system('cp -r ' + subfolder + '/. ' + emotion_folder_dict.get(tess_emotions.get(subfolder.split('_')[len(subfolder.split('_')) - 1])))

# Check Distribution and Process Data

In [5]:
# get emotion categories
path = 'drive/MyDrive/processed_emotion_dataset/'
sub_folder = glob.glob(path + '*')
emotions = [i.split('/')[3] for i in sub_folder]
emotions

['neutral', 'calm', 'happy', 'sad', 'angry', 'fearful', 'disgust', 'surprised']

In [45]:
# print count of each emotion category
for cat in emotions:
  f_list = glob.glob(path + cat + '/*')
  print('Emotion category: \'{}\' Count: {}'.format(cat,len(f_list)))

Emotion category: 'neutral' Count: 1619
Emotion category: 'happy' Count: 1739
Emotion category: 'sad' Count: 1739
Emotion category: 'angry' Count: 1739
Emotion category: 'fearful' Count: 1739
Emotion category: 'disgust' Count: 1739
Emotion category: 'surprised' Count: 469


In [ ]:
# group calm and neutral
source = sub_folder[1] + '/.'
target = sub_folder[0] + '/.'
os.system('cp -r ' + source + ' ' + target)

0

In [6]:
# remove calm folder from list and count of each emotion category
sub_folder.remove(sub_folder[1])
emotions = [i.split('/')[3] for i in sub_folder]
for cat in emotions:
  f_list = glob.glob(path + cat + '/*')
  print('Emotion category: \'{}\' Count: {}'.format(cat,len(f_list)))

Emotion category: 'neutral' Count: 1619
Emotion category: 'happy' Count: 1739
Emotion category: 'sad' Count: 1739
Emotion category: 'angry' Count: 1739
Emotion category: 'fearful' Count: 1739
Emotion category: 'disgust' Count: 1739
Emotion category: 'surprised' Count: 468


In [ ]:
# get max and min mel frequency array for padding
ms_max = (0,0)
for cat in sub_folder:
  f_list = glob.glob(cat + '/*')
  for file_name in f_list:
    x, sr = librosa.load(file_name)
    melspect = librosa.feature.melspectrogram(y=x, sr=sr)
    if melspect.shape > ms_max:
      ms_max = melspect.shape

print(ms_max)

In [48]:
for cat, sub_fold in zip(emotions, sub_folder):
  folder = sub_fold + "/"
  emotion_files = glob.glob(folder + "*")
  num = 0
  for f in emotion_files:
    extension = f.split(".")[-1]
    new_f_name = cat + "_" + str(num) + "." + extension
    os.rename(f, folder + new_f_name)
    num += 1
  num = 0

In [ ]:
x, sr = librosa.load(sub_folder[6] + '/' + f_name)
# stft = librosa.feature.chroma_stft(y=x, sr=sr)
# print(stft.shape)
# cqt = librosa.feature.chroma_cqt(y=x, sr=sr)
# print(cqt.shape)
n_mels = 128
melspect = librosa.feature.melspectrogram(y=x, sr=sr, n_mels=n_mels, pad_mode='constant')
# pad_val = n_mels-melspect.shape[1]
pad_val = 300
melspect = np.pad(melspect, [(0,0),(0,pad_val)], mode='constant')
print(melspect.shape)
mfcc = librosa.feature.mfcc(y=x, sr=sr)
print(mfcc.shape)
spec_roll = librosa.feature.spectral_rolloff(y=x, sr=sr)
print(spec_roll.shape)


(128, 379)
(20, 79)
(1, 79)


In [ ]:
melspect[:,0:128].shape

(128, 128)

In [ ]:
# rename each file


# Create Custom Dataset

Emotion category: 'neutral' Count: 1619
Emotion category: 'happy' Count: 1739
Emotion category: 'sad' Count: 1739
Emotion category: 'angry' Count: 1739
Emotion category: 'fearful' Count: 1739
Emotion category: 'disgust' Count: 1739
Emotion category: 'surprised' Count: 468

In [ ]:
sub_folder

['drive/MyDrive/processed_emotion_dataset/neutral',
 'drive/MyDrive/processed_emotion_dataset/happy',
 'drive/MyDrive/processed_emotion_dataset/sad',
 'drive/MyDrive/processed_emotion_dataset/angry',
 'drive/MyDrive/processed_emotion_dataset/fearful',
 'drive/MyDrive/processed_emotion_dataset/disgust',
 'drive/MyDrive/processed_emotion_dataset/surprised']

In [ ]:
class audiodata(Dataset):
  def __init__(self, file_path, n_mels=128):
    self.mels = n_mels
    self.file_path = file_path
    self.data = []
    for sub_path in file_path:
      emotion_category = sub_path.split("/")[-1]
      for audio_path in glob.glob(sub_path + "/*"):
        self.data.append([audio_path, emotion_category])
    self.class_dict = {"neutral": 0,
                    "happy": 1,
                    "sad": 2,
                    "angry": 3,
                    "fearful": 4,
                    "disgust": 5,
                    "surprised": 6 
                    }

  def __len__(self):
    return len(self.data)

  def __getitem__(self,index):
    audio_path, emotion_class = self.data[index]
    x, sr = librosa.load(audio_path)
    melspect = librosa.feature.melspectrogram(y=x, sr=sr, n_mels=128, pad_mode='constant')
    pad_crop_length = 128 - melspect.shape[1]
    if pad_crop_length > 0:
      melspect = np.pad(melspect, [(0,0),(0,pad_crop_length)], mode='constant')
    if pad_crop_length < 0:
      melspect = melspect[:,0:128]
    # get emotion class code
    class_code = self.class_dict[emotion_class]
    audio_tensor = torch.from_numpy(melspect)
    # audio_tensor = audio_tensor.permute(2, 0, 1)
    return audio_tensor, class_code